### PR TITLE
Use standard pyproject.toml

### LINKED ISSUE #56: Consider a more standard build tool?
May I ask why you chose poetry as your build tool? Would you consider transitioning to something more standard (poetry uses its own non-standard fields in `pyproject.toml`)? I think the transition should be pretty straightforward, and then any standard-compliant tool could be used, including [uv](https://docs.astral.sh/uv/) (which should have almost all features and is extremely fast).

### PR DESCRIPTION
**Pull request by cbrnr**
Fixes #56.

### CODE REVIEW

**Comment by hofaflo** [comment]
Thanks for the PR! Are you already working on getting the checks to pass or should I look into it?

**Comment by cbrnr** [comment]
Thanks @hofaflo. I tried to modify `test.yml`, but I'm not sure what you are doing with NumPy. Do you want to install the specific versions depending on the Python version? If so, I'm not sure why it's not working (`${{matrix.numpy}}` yields `--pre numpy`, but I thought it should be e.g. `numpy==1.22.0`)...

**Comment by cbrnr** [comment]
So what I'm saying is that I don't know how to specify using a pre-release NumPy...

**Comment by cbrnr** [comment]
I guess this will require using the `uv pip` interface, which will probably make the process more involved.

**Comment by cbrnr** [comment]
@hofaflo a test is failing because of a user warning. Did I miss something? I don't think this is related to the changes in this PR.

**Comment by cbrnr** [comment]
Regarding building the docs, I'm not sure how you would like to declare `edfio` as a dependency, any suggestions?

**Comment by hofaflo** [comment]
> I'm not sure what you are doing with NumPy. Do you want to install the specific versions depending on the Python version?

We're testing with the oldest and newest supported NumPy version for each Python version. If I understand correctly, for `uv` we'd have to replace `numpy: ["--pre numpy"]` with `numpy: ["numpy --prerelease=allow"]`. However, this flag was introduced to ensure compatibility with NumPy v2.0 before there was a non-pre-release, so we can just do `numpy: ["numpy"]` now.


> a test is failing because of a user warning. Did I miss something? I don't think this is related to the changes in this PR.

This is probably because `pytest.warns` re-emits unmatched warnings since v8.0.0, which was not used so far as we pinned the dev dependencies to exact versions (something I'd like to keep doing). 


> Regarding building the docs, I'm not sure how you would like to declare `edfio` as a dependency, any suggestions?

Simplifying [this example](https://docs.readthedocs.io/en/stable/build-customization.html#install-dependencies-with-uv) from the RTD docs a bit, maybe something like

```yml
  commands:
    - pip install uv
    - uv run --extra dev python -m sphinx -T -b html -d docs/_build/doctrees -D language=en docs/source $READTHEDOCS_OUTPUT/html
```

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -41,25 +41,15 @@ jobs:
         with:
```
Comment (hofaflo) [inline .github/workflows/test.yml:41]: I think the `setup-python` action is not needed anymore if we use `setup-uv` (see [here](https://github.com/astral-sh/setup-uv?tab=readme-ov-file#do-i-still-need-actionssetup-python-alongside-setup-uv)).

**Comment by cbrnr** [comment]
Never mind, I found it.

**Comment by cbrnr** [comment]
Do you want to bump any dev packages while I'm at it @hofaflo?

**Comment by hofaflo** [comment]
> Do you want to bump any dev packages while I'm at it?

No thanks, I think that would fit better in a separate PR.

Could you update the section "Development environment" accordingly? E.g. to something like this:

```md
We suggest to use [`uv`](https://docs.astral.sh/uv/getting-started/installation/) for managing project dependencies, but any other standard-compliant tool will work too.

Create a virtual environment with Python 3.9 and install all dependencies:

    uv sync --python=3.9 --all-extras

**When using `uv`, prepend the commands below with `uv run` to make sure they are executed in the virtual environment!**
```

We'll need to replace `poetry` in the release pipeline as well. I just added this repo as a trusted publisher on PyPI, so the following setup should work (in `release.yml`):

```yml
  publish:
    needs: test
    runs-on: ubuntu-latest
    permissions:
      id-token: write
    steps:
      - uses: actions/checkout@v4
        with:
          lfs: true
      - name: Setup uv
        uses: astral-sh/setup-uv@v3
      - name: Publish to PyPI
        run: |
          uv build
          uv publish
```

Finally, we should remove remaining mentions of `poetry` in `README.md` and `CONTRIBUTING.md`

**Comment by cbrnr** [comment]
Great, will do!

Finally, do you think we can skip

```
      - name: Install Python ${{ matrix.python-version }}
        run: uv python install ${{ matrix.python-version }}
```

and instead use

```
      - name: Pre-commit
        run: |
          uv run --extra dev --python ${{ matrix.python-version }} --with ${{ matrix.numpy }} pre-commit run --all-files
      - name: Pytest
        run: |
          uv run --extra dev --python ${{ matrix.python-version }} --with ${{ matrix.numpy }} pytest -Werror --cov --cov-report term-missing --cov-fail-under=100
```

?

**Comment by cbrnr** [comment]
And another question/heads up regarding the dynamic version. If I install the current `HEAD`, the version shown is `0.4.0.dev94+g47f9643`. Is this OK? I'm not sure, because the latest release/tag is `v0.4.4`, so maybe the dev version should be `0.5.0.dev???????`?

**Comment by cbrnr** [comment]
PS: I'm not even sure what's going on, because `uv tool run hatch version` outputs `0.5.0.dev20+g47f9643`, so maybe everything is fine?

**Comment by cbrnr** [comment]
The dynamic version is incorrect for editable installs unless the project is rebuilt (https://github.com/ofek/hatch-vcs?tab=readme-ov-file#editable-installs), so all good! However, what do you want the dev version to look like?

1. Bump minor tag (e.g., `0.5.0.dev20+g47f9643`)
2. Bump patch tag (e.g., `0.4.5.dev20+g47f9643.d20241014`, the default)

**Comment by hofaflo** [comment]
> Great, will do!
> 
> Finally, do you think we can skip
> 
> ```
>       - name: Install Python ${{ matrix.python-version }}
>         run: uv python install ${{ matrix.python-version }}
> ```
> 
> and instead use
> 
> ```
>       - name: Pre-commit
>         run: |
>           uv run --extra dev --python ${{ matrix.python-version }} --with ${{ matrix.numpy }} pre-commit run --all-files
>       - name: Pytest
>         run: |
>           uv run --extra dev --python ${{ matrix.python-version }} --with ${{ matrix.numpy }} pytest -Werror --cov --cov-report term-missing --cov-fail-under=100
> ```
> 
> ?

I'd like to leave this as-is to keep the step commands shorter.


> what do you want the dev version to look like?
> 
>   1. Bump minor tag (e.g., `0.5.0.dev20+g47f9643`, currently implemented in this PR)
> 
>   2. Bump patch tag (e.g., `0.4.5.dev20+g47f9643.d20241014`, the default)

Thanks for bringing this up! I think the default is more appropriate, as we aren't using release branches anyway.

**Comment by cbrnr** [comment]
OK, I've switched to the default versioning scheme, I think this is now ready for review!

**Review by hofaflo — APPROVED** [review]
Looks perfect, thanks @cbrnr!

**Comment by cbrnr** [comment]
Fingers crossed that everything works in the release workflow! But I guess we'll find out soon...

**Comment by cbrnr** [comment]
(Or maybe you test on test-pypi before the next release.)

**Pull request merged**